### PR TITLE
8233343: Deprecate -XX:+CriticalJNINatives flag which implements JavaCritical native functions

### PR DESCRIPTION
**Pull request by coleenp**
This change deprecates the -XX:+CriticalJNINatives flag and removes the develop flag -XX:+StressCriticalJNINatives.  See CSR for more details.

This change also removes the lazy GC lock in the critical native transition, and runs the critical native function as thread_in_Java.  I add a safepoint check at the end of the native function and transition to native and poll again for the safepoint after the function if a safepoint is requested.

Tested with tier 1-6 (we have a few tests that use this).  And built on linux-x86-open,linux-s390x-open,linux-arm32-debug,linux-ppc64le-debug.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ❌ (1/9 failed) | ✔️ (9/9 passed) |

**Failed test task**
- [Windows x64 (hs/tier1 compiler)](https://github.com/coleenp/jdk/runs/1286847967)

### Issue
 * [JDK-8233343](https://bugs.openjdk.java.net/browse/JDK-8233343): Deprecate -XX:+CriticalJNINatives flag which implements JavaCritical native functions

### Reviewers
 * [Robbin Ehn](https://openjdk.java.net/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to fa23ec199cd8a532e534314fc10619dff415814d
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Zhengyu Gu](https://openjdk.java.net/census#zgu) (@zhengyu123 - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/764/head:pull/764`
`$ git checkout pull/764`

### CODE REVIEW

**Review by robehn — APPROVED** [review]
Thanks you @coleenp!
Looks good, really nice delta of -800 LOC !

**Comment by TheRealMDoerr** [comment]
Makes sense to me. But I have a couple of remarks/suggestions:
- Object pinning for T_ARRAY on x86 shouldn't be needed any more since we stay in _thread_in_Java
- Transition to _thread_in_native is pointless if we transition to _thread_in_native_trans immediately afterwards
- I think the tests should also run on os.arch=="ppc64" | os.arch=="ppc64le" | os.arch=="s390x", but we should double-check if they really work
Thanks for taking care of all platforms!

**Comment by coleenp** [comment]
@TheRealMDoerr Thank you for reviewing this and your comments.

- yes, I can remove object pinning since it's no longer needed.
- The transition to native is pointless but I was a bit unnerved from going from _thread_in_Java to _thread_in_native_trans even though there's no consistency checks to show it's an illegal transition.  On the other hand, the dummy transition to native isn't nice either so I'll make this change and retest.
- I can't test with ppc or s390 but can you modify the requires in the two tests in this patch and let me know if it works?

**Comment by zhengyu123** [comment]
> @TheRealMDoerr Thank you for reviewing this and your comments.
> 
> * yes, I can remove object pinning since it's no longer needed.
> * The transition to native is pointless but I was a bit unnerved from going from _thread_in_Java to _thread_in_native_trans even though there's no consistency checks to show it's an illegal transition.  On the other hand, the dummy transition to native isn't nice either so I'll make this change and retest.
> * I can't test with ppc or s390 but can you modify the requires in the two tests in this patch and let me know if it works?

Agree, object pinning is no longer needed for Shenandoah.

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp [hunk 1]
```diff
@@ -2352,17 +2184,19 @@ nmethod *SharedRuntime::generate_native_wrapper(MacroAssembler *masm,
...
-  // Publish thread state
-  // --------------------------------------------------------------------------
+  if (!is_critical_native) {
+    // Publish thread state
+    // --------------------------------------------------------------------------
 
-  // Use that pc we placed in r_return_pc a while back as the current frame anchor.
-  __ set_last_Java_frame(R1_SP, r_return_pc);
+    // Use that pc we placed in r_return_pc a while back as the current frame anchor.
+    __ set_last_Java_frame(R1_SP, r_return_pc);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:2192]: These 2 lines need to get moved before if (!is_critical_native) check.
With this change, the 2 tests have passed with " | os.arch=="ppc64" | os.arch=="ppc64le" | os.arch=="s390x"" added.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:2192]: Ok, I moved those lines and added the other platforms for the basic CriticalNative tests that I modified.  Thank you so much for testing this on the other platforms.

**Review by TheRealMDoerr — APPROVED** [review]
Looks good. Thanks!

**Review by zhengyu123 — APPROVED** [review]

**Comment by coleenp** [comment]
Thank you for the code reviews!
/integrate

**Pull request closed without merging**